experimental/casemap/src/closer.rs

```diff
@@ -0,0 +1,226 @@
+/// let found = cm.add_string_case_closure("ss", &mut builder);
+/// assert!(found);
+/// let set = builder.build();
+///
+/// assert!(set.contains('ß'));
+/// assert!(set.contains('ẞ'));
+/// ```
+///
+/// <div class="stab unstable">
+/// 🚧 This code is experimental; it may change at any time, in breaking or non-breaking ways,
+/// including in SemVer minor releases. It can be enabled with the "experimental" Cargo feature
+/// of the icu meta-crate. Use with caution.
+/// <a href="https://github.com/unicode-org/icu4x/issues/2535">#2535</a>
+/// </div>
+#[derive(Clone, Debug)]
+pub struct CaseMapCloser<CM> {
+    cm: CM,
+    unfold: DataPayload<CaseMapUnfoldV1Marker>,
+}
+
+#[cfg(feature = "compiled_data")]
+impl Default for CaseMapCloser<CaseMapper> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CaseMapCloser<CaseMapper> {
+    /// A constructor which creates a [`CaseMapCloser`].
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use icu_casemap::CaseMapCloser;
+    /// use icu_collections::codepointinvlist::CodePointInversionListBuilder;
+    ///
+    /// let cm = CaseMapCloser::new();
+    /// let mut builder = CodePointInversionListBuilder::new();
+    /// let found = cm.add_string_case_closure("ffi", &mut builder);
+    /// assert!(found);
+    /// let set = builder.build();
+    ///
+    /// assert!(set.contains('ﬃ'));
+    ///
+    /// let mut builder = CodePointInversionListBuilder::new();
+    /// let found = cm.add_string_case_closure("ss", &mut builder);
+    /// assert!(found);
+    /// let set = builder.build();
+    ///
+    /// assert!(set.contains('ß'));
+    /// assert!(set.contains('ẞ'));
+    /// ```
+    ///
+    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    #[cfg(feature = "compiled_data")]
+    pub const fn new() -> Self {
+        Self {
+            cm: CaseMapper::new(),
+            unfold: DataPayload::from_static_ref(
+                crate::provider::Baked::SINGLETON_PROPS_CASEMAP_UNFOLD_V1,
+            ),
+        }
+    }
+
+    icu_provider::gen_any_buffer_data_constructors!(locale: skip, options: skip, error: DataError,
+    #[cfg(skip)]
+    functions: [
+        new,
+        try_new_with_any_provider,
+        try_new_with_buffer_provider,
+        try_new_unstable,
+        Self,
+    ]);
+
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(UNSTABLE, Self::new)]
+    pub fn try_new_unstable<P>(provider: &P) -> Result<Self, DataError>
+    where
+        P: DataProvider<CaseMapV1Marker> + DataProvider<CaseMapUnfoldV1Marker> + ?Sized,
+    {
+        let cm = CaseMapper::try_new_unstable(provider)?;
+        let unfold = provider.load(Default::default())?.take_payload()?;
+        Ok(Self { cm, unfold })
+    }
+}
+
+// We use Borrow, not AsRef, since we want the blanket impl on T
+impl<CM: Borrow<CaseMapper>> CaseMapCloser<CM> {
```

Comment:
> Suggestion: maybe `AsRef<CaseMapperBorrowed>` and then we impl that on `CaseMapper`. (Or did we decide not to have `CaseMapperBorrowed`?)

We decided not to, but furthermore, it won't work: `AsRef<CMB>` would need to return an `&CMB` and we don't have a place to anchor that referemce.


> The `Borrow` trait is a bit restricting and I prefer `AsRef` unless you really need the Eq/Ord invariants of `Borrow`, which we don't need here. Just implement `AsRef` where you need it.

I've basically never found this to be a problem because 99% of the time you're using some stdlib smart pointer and those work just fine. 

But I'll add the manual impl and switch over, that ought to work.